crates/trigger-http/src/server.rs

```diff
@@ -138,16 +148,31 @@ impl<F: RuntimeFactors> HttpServer<F> {
 
     /// Serve incoming requests over the provided [`TcpListener`].
     pub async fn serve(self: Arc<Self>) -> anyhow::Result<()> {
-        let listener = TcpListener::bind(self.listen_addr).await.with_context(|| {
-            format!(
-                "Unable to listen on {listen_addr}",
-                listen_addr = self.listen_addr
-            )
-        })?;
+        let listener: anyhow::Result<TcpListener> = match TcpListener::bind(self.listen_addr).await
+        {
+            Ok(listener) => Ok(listener),
+            Err(err) => {
+                if err.kind() == ErrorKind::AddrInUse {
+                    let mut addr = self.listen_addr;
+                    addr.set_port(self.port.unwrap_or(0));
+                    let listener = TcpListener::bind(addr).await.with_context(|| {
+                        format!(
+                            "Unable to listen on {listen_addr}",
+                            listen_addr = self.listen_addr
+                        )
+                    })?;
+
+                    Ok(listener)
+                } else {
+                    Err(anyhow::anyhow!("Unable to listen on {}", self.listen_addr))
+                }
+            }
+        };
+
         if let Some(tls_config) = self.tls_config.clone() {
-            self.serve_https(listener, tls_config).await?;
+            self.serve_https(listener?, tls_config).await?;
```

Comment:
I'm not sure I love having to `?` the listener everywhere it's used.  Why not stick with the previous pattern of failing errors _before_ assigning the `listener` variable?